pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -3764,6 +3709,164 @@ func TestApplyTaskResultsToPipelineResults(t *testing.T) {
+		}},
+		taskResults: map[string][]v1beta1.TaskRunResult{
+			"pt1": {
+				{
+					Name:  "foo",
+					Value: *v1beta1.NewStructuredValues("do"),
+				}, {
+					Name:  "bar",
+					Value: *v1beta1.NewStructuredValues("mi"),
+				},
+			},
+			"pt2": {{
+				Name:  "baz",
+				Value: *v1beta1.NewStructuredValues("rae"),
+			}},
+		},
+		expectedResults: []v1beta1.PipelineRunResult{{
+			Name:  "pipeline-result-1",
+			Value: *v1beta1.NewStructuredValues("do"),
+		}, {
+			Name:  "pipeline-result-2",
+			Value: *v1beta1.NewStructuredValues("do, rae, mi, rae, do"),
+		}},
+	}, {
+		description: "multiple-results-custom-and-normal-tasks",
+		results: []v1beta1.PipelineResult{{
+			Name:  "pipeline-result-1",
+			Value: *v1beta1.NewStructuredValues("$(tasks.customtask.results.foo)"),
+		}, {
+			Name:  "pipeline-result-2",
+			Value: *v1beta1.NewStructuredValues("$(tasks.customtask.results.foo), $(tasks.normaltask.results.baz), $(tasks.customtask.results.bar), $(tasks.normaltask.results.baz), $(tasks.customtask.results.foo)"),
+		}},
+		runResults: map[string][]v1beta1.CustomRunResult{
+			"customtask": {
+				{
+					Name:  "foo",
+					Value: "do",
+				}, {
+					Name:  "bar",
+					Value: "mi",
+				},
+			},
+		},
+		taskResults: map[string][]v1beta1.TaskRunResult{
+			"normaltask": {{
+				Name:  "baz",
+				Value: *v1beta1.NewStructuredValues("rae"),
+			}},
+		},
+		expectedResults: []v1beta1.PipelineRunResult{{
+			Name:  "pipeline-result-1",
+			Value: *v1beta1.NewStructuredValues("do"),
+		}, {
+			Name:  "pipeline-result-2",
+			Value: *v1beta1.NewStructuredValues("do, rae, mi, rae, do"),
+		}},
+	}, {
+		description: "multiple-results-skipped-and-normal-tasks",
+		results: []v1beta1.PipelineResult{{
+			Name:  "pipeline-result-1",
+			Value: *v1beta1.NewStructuredValues("$(tasks.skippedTask.results.foo)"),
+		}, {
+			Name:  "pipeline-result-2",
+			Value: *v1beta1.NewStructuredValues("$(tasks.skippedTask.results.foo), $(tasks.normaltask.results.baz)"),
+		}},
+		taskResults: map[string][]v1beta1.TaskRunResult{
+			"normaltask": {{
+				Name:  "baz",
+				Value: *v1beta1.NewStructuredValues("rae"),
+			}},
+		},
+		skippedTasks: []v1beta1.SkippedTask{{
+			Name: "skippedTask",
+		}},
+		expectedResults: nil,
+	}} {
+		t.Run(tc.description, func(t *testing.T) {
+			received, err := ApplyTaskResultsToPipelineResults(context.Background(), tc.results, tc.taskResults, tc.runResults, tc.skippedTasks)
+			if err != nil {
+				t.Errorf("Got unecpected error:%v", err)
+			}
+			if d := cmp.Diff(tc.expectedResults, received); d != "" {
+				t.Errorf(diff.PrintWantGot(d))
+			}
+		})
+	}
+}
+
+func TestApplyTaskResultsToPipelineResults_Error(t *testing.T) {
+	for _, tc := range []struct {
+		description     string
+		results         []v1beta1.PipelineResult
+		taskResults     map[string][]v1beta1.TaskRunResult
+		runResults      map[string][]v1beta1.CustomRunResult
+		expectedResults []v1beta1.PipelineRunResult
```

Comment:
expectedResults is always nil right? Can we remove it and simplify the cmp.Diff below for the error tests?

Reply:
Yeah I have thought about it, I think if we keep it we could make sure that for error cases we don't emit pipeline results.

Reply:
emmm looks like there is a case "mixed-success-tasks-some-returned-results" is returning results. 🤔

Reply:
It maybe a valid case when pipelinerun failed but still emit results: https://github.com/tektoncd/pipeline/issues/3749

Reply:
ah, I missed that one